src/main/scala/coupledL2/tl2chi/MSHR.scala

```diff
@@ -829,6 +829,16 @@ class MSHR(implicit p: Parameters) extends TL2CHIL2Module {
 
   // RXDAT
   when (rxdat.valid) {
+    when (rxdat.bits.chiOpcode.get === DataSepResp) {
+      require(beatSize == 2) // TODO: This is ugly
+      state.w_grantfirst := true.B
+      state.w_grantlast := state.w_grantfirst
+      state.w_grant := req.off === 0.U || state.w_grantfirst  // TODO? why offset?
```

Comment:
The CompAck must only be sent after a CompData or RespSepData (instead of DataSepResp) is received. Therefore `w_grant.*` should be set after RespSepData is received, otherwise CompAck might be sent before RespSepData if RespSepData arrives in RN after DataSepResp.

Also, extra logics besides `w_grant.*` are necessary to make sure that Grant upwards is sent only after both response and data are received.